src/functional_simulator.cpp

```
@@ -192,8 +197,48 @@ void FunctionalSimulator::execute() {
     }
 }
 
+/**
+ * @brief Performs memory operations for the instruction in the MEMORY stage.
+ *
+ * This method handles data memory access for load and store instructions.
+ * For load instructions (LDW), it reads memory at the computed address and
+ * stores the value in the pipeline data for use in the write-back stage.
+ * For store instructions (STW), it writes the value from the rt register
+ * to the computed memory address and logs the access in the stats.
+ */
 void FunctionalSimulator::memory() {
-    // Stub: Access memory if needed (load or store only)
+    // Get reference to PipelineStageData pointer in memory stage
+    // using auto to automatically deduce the type
+    auto& mem_data = pipeline[MEMORY];
+
+    // In the case of a stall cycle or no instruction in
```

This all looks good but the comment is sort of confusing. How are stall cycles coming into play here?

---

This is mainly noting that it could be valid to have a null item in this stage, if a bubble had been inserted for a stall earlier. If it weren't valid, I'd want to throw an error or something rather than just return.

---

That makes sense, I was thinking about this wrong.